### PR TITLE
Extend active students amount logic, calendar periods and cache logic 

### PR DESCRIPTION
**Pull request by DmytryiStriletskyi**
Pull request contains extended active students amount logic, calendar periods (day, week, month) and cache logic, also documentation.

### CODE REVIEW

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 1]
```diff
@@ -19,6 +22,132 @@
 from .models import TokenStorage
 
 
+UTC = pytz.UTC
+
+
+def calculate_active_students_amount(query_name, activity_period):
+    """
+    Calculates instance active students amount for particular period as like as previous calendar day,
+    week and month after cached.
```
Comment (auraz) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Follow edx doc string style, (empty line after 1st string missed)

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 2]
```diff
@@ -19,6 +22,132 @@
...
+        end_of_month (datetime): Previous month`s last day end. Example for may is 2017-04-30 23:59:59.
+    """
+
+    last_day_of_previous_month = (datetime.datetime.now().replace(day=1) - datetime.timedelta(days=1)).replace(
+        tzinfo=UTC
+    )
+
+    _, days_in_month = monthrange(last_day_of_previous_month.year, last_day_of_previous_month.month)
+
+    start_of_month = (last_day_of_previous_month - datetime.timedelta(days=days_in_month-1)).replace(
```
Comment (auraz) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: more elegant https://gist.github.com/waynemoore/1109153

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 3]
```diff
@@ -19,6 +22,132 @@
...
+    """
+    Calculates instance students per country amount for particular period as like as previous calendar day,
+    week and month after cached.
+
+    Returns:
+        students_per_country_after_cached (dict): Country-count accordance as pair of key-value.
+                                                  Example: {u'FR': 3434, u'UA': 1124, None: 2345}
+    """
+
+    period_start, period_end = activity_period
```
Comment (auraz) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Are you sure that you can't measure periods in days?

Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: I researched Django`s core logic, I can. Thanks!

File: openedx/core/djangoapps/edx_global_analytics/tasks.py (hunk 2 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Thanks!

File: openedx/core/djangoapps/edx_global_analytics/tasks.py (hunk 3 above)
Comment (auraz) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Great, you're welcome!

File: openedx/core/djangoapps/edx_global_analytics/utils.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+    Returns:
+        start_of_month (date): Calendar month`s start day. Example for may is 2017-04-01.
+        end_of_month (date): Calendar month`s end day, it`s the first day of next month, that doesn't count in segment.
+                             Example for may is 2017-05-01.
+    """
+    previous_month_date = datetime.date.today().replace(day=1) - datetime.timedelta(days=1)
+
+    start_of_month = previous_month_date.replace(day=1)
+    end_of_month = previous_month_date.replace(
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: Logic here should be like this: `end_of_month = start_of_month + timedelta(days=days_in_prev_month-1)`
Current implementation looks error-prone.

File: openedx/core/djangoapps/edx_global_analytics/utils.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+    Returns:
+        students_per_country_after_cached (dict): Country-count accordance as pair of key-value.
+                                                  Example: {u'FR': 3434, u'UA': 1124}
+    """
+    period_start, period_end = activity_period
+
+    students_per_country = dict(UserProfile.objects.exclude(
+        Q(user__last_login=None) | Q(user__is_active=False)
+    ).filter(user__last_login__gt=period_start, user__last_login__lt=period_end
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: Same thing about `__gte` & `__lte`

File: openedx/core/djangoapps/edx_global_analytics/utils.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+import calendar
+
+from django.core.cache import cache
+
+from django.db.models import Count
+from django.db.models import Q
+from student.models import UserProfile
+
+
+def calculate_active_students_amount(query_name, activity_period):
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: Also you can use memoize pattern here see [django-memoize](https://pythonhosted.org/django-memoize/) for the reference

Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: I`ll fix it.

File: openedx/core/djangoapps/edx_global_analytics/utils.py (hunk 5 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: I researched Django`s logic for that, I`ll implement `__gte` feature, it works fine.

File: openedx/core/djangoapps/edx_global_analytics/utils.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    """
+    Calculates instance active students amount for particular period as like as previous calendar day,
+    week and month after cached.
+    """
+    period_start, period_end = activity_period
+
+    active_students_amount = UserProfile.objects.exclude(
+        Q(user__last_login=None) | Q(user__is_active=False)
+    ).filter(user__last_login__gt=period_start, user__last_login__lt=period_end).count()
```
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: I researched Djangos logic for that, Ill implement __gte feature, it works fine.

File: openedx/core/djangoapps/edx_global_analytics/utils.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+    Returns:
+        start_of_month (date): Calendar week`s start day. Example for may is 2017-05-08.
+        end_of_month (date): Calendar week`s end day, it`s the first day of next week, that doesn't count in segment.
+                             Example for may is 2017-05-15.
+    """
+    days_after_week_started = datetime.date.today().weekday() + 7
+
+    start_of_week = datetime.date.today() - datetime.timedelta(days=days_after_week_started)
+    end_of_week = start_of_week + datetime.timedelta(days=7)
```
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: Discussed.

File: openedx/core/djangoapps/edx_global_analytics/utils.py (hunk 4 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: Discussed.

File: openedx/core/djangoapps/edx_global_analytics/utils.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+    start_of_month = previous_month_date.replace(day=1)
+    end_of_month = previous_month_date.replace(
+        day=calendar.monthrange(previous_month_date.year, previous_month_date.month)[1]
+    ) + datetime.timedelta(days=1)
+
+    return start_of_month, end_of_month
+
+
+def caching_instance_data(query_name, query, clear_cache_date):
```
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: I will remove this from cache logic or rename it, thanks.

File: openedx/core/djangoapps/edx_global_analytics/utils.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    """
+    if datetime.date.today() == clear_cache_date:
+        cache.delete(query_name)
+
+    cached_query = cache.get(query_name)
+
+    if cached_query is not None:
+        return cached_query
+
+    cache.set(query_name, query)
```
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: Thanks.

File: openedx/core/djangoapps/edx_global_analytics/utils.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    including long time unchangeable as like as monthly statistics.
+
+    Arguments:
+        query_name (str): Name of query.
+        query (query result): Django-query.
+        clear_cache_date (date): Last date of saving cache data.
+
+    Returns cached query result.
+    """
+    if datetime.date.today() == clear_cache_date:
```
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: I will use cache`s timeout feature instead. Thanks.

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 12]
```diff
@@ -3,21 +3,91 @@
...
 from django.conf import settings
 from django.contrib.sites.models import Site
-from django.db.models import Count
-from django.db.models import Q
 
 from xmodule.modulestore.django import modulestore
-from student.models import UserProfile
 from .models import TokenStorage
 
+from .utils import calculate_active_students_amount, calculate_students_per_country, \
```
Comment (TamoshaytisV) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: please refactor to look like the following:

```
from .utils import (
    calculate_active_students_amount,
    calculate_students_per_country,
    ...
)
```

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 13]
```diff
@@ -3,21 +3,91 @@
...
+                    get_previous_day_start_and_end_dates, get_previous_week_start_and_end_dates, \
+                    get_previous_month_start_and_end_dates
+
+
+def calculate_instance_information():
```
Comment (TamoshaytisV) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: I would suggest renaming this method to smth like `get_instance_information` or `fetch_instance_information`

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 14]
```diff
@@ -42,22 +112,12 @@ def count_data():
...
-    ).annotate(count=Count('country'))
-
-    # Get courses amount within current platform.
-    courses_amount = len(modulestore().get_courses())
+    # Get all information, that need to be calculated.
+    (active_students_amount_day,
+     active_students_amount_week,
+     active_students_amount_month,
+     courses_amount,
+     students_per_country) = calculate_instance_information()
```
Comment (TamoshaytisV) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Why not to call `calculate_instance_information` in SEND method and avoid transferring bunch of parameters?

**Comment by DmytryiStriletskyi** [comment]
I found we sent statistics data per level after we got it all, that was a fail.
Now at first we get statistic level and then we know - do we need to get fron DB a particular bunch of data or not.

File: openedx/core/djangoapps/edx_global_analytics/tasks.py (hunk 12 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Done.

File: openedx/core/djangoapps/edx_global_analytics/tasks.py (hunk 13 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Done. Thanks!

File: openedx/core/djangoapps/edx_global_analytics/tasks.py (hunk 14 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Not actual after refactor. Thanks.

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+from student.tests.factories import UserFactory
+
+from ..utils import fetch_instance_information, cache_instance_data
+
+
+class TestStudentsAmountPerParticularPeriod(TestCase):
+    """
+    Tests cover all methods, that have a deal with statistics calculation.
+    """
+
```
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: I will fix unnecessary indent in next commits with Jitter implementation. Don`t pay attention to this.

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 16]
```diff
@@ -135,4 +139,9 @@ def count_data():
             'students_per_country': json.dumps(students_per_country)
         })
 
-    requests.post(post_url, data)
+    try:
+        request = requests.post(post_url, data)
+        logger.info('Connected without error to {0}'.format(request.url))
+        logger.info('Data was successfully sent. Status code is {0}.'.format(request.status_code))
+    except requests.ConnectionError as error:
```
Comment (TamoshaytisV) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Please consider changing `ConnectionError` to `RequestException `

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 17]
```diff
@@ -3,21 +3,91 @@
...
+        data.update({
+            'latitude': latitude,
+            'longitude': longitude,
+            'platform_name': platform_name,
+            'platform_url': platform_url,
+            'statistics_level': 'enthusiast',
+            'students_per_country': json.dumps(students_per_country)
+        })
+
+    requests.post(post_url, data)
```
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Done. Jitter discussed.

File: openedx/core/djangoapps/edx_global_analytics/tasks.py (hunk 16 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: @TamoshaytisV, log in terminal shows me `ConnectionError` exception few times, what is reason to follow `RequestException`?

Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: @TamoshaytisV, log in terminal shows me `ConnectionError` exception few times, what is reason to follow `RequestException`?

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 18]
```diff
@@ -80,31 +89,59 @@ def count_data():
...
+        platform_latitude = olga_settings.get("PLATFORM_LATITUDE")
+        platform_longitude = olga_settings.get("PLATFORM_LONGITUDE")
+
+        if platform_latitude and platform_longitude:
+            latitude, longitude = platform_latitude, platform_longitude
+        else:
+            latitude, longitude = ip_data_json['latitude'], ip_data_json['longitude']
+
+        # Platform name.
+        if settings.PLATFORM_NAME:
```
Comment (maxsocl) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: @DmytryiStriletskyi 
Please consider to change this to something like following:
```
platform_name = settings.PLATFORM_NAME or Site.objects.get_current()
```

File: openedx/core/djangoapps/edx_global_analytics/tasks.py (hunk 16 above)
Comment (TamoshaytisV) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: please read http://docs.python-requests.org/en/latest/user/quickstart/#errors-and-exceptions
```
In the event of a network problem (e.g. DNS failure, refused connection, etc), Requests will raise a ConnectionError exception.
In the event of the rare invalid HTTP response, Requests will raise an HTTPError exception.
If a request times out, a Timeout exception is raised.
If a request exceeds the configured number of maximum redirections, a TooManyRedirects exception is raised.
All exceptions that Requests explicitly raises inherit from requests.exceptions.RequestException.
```
You can either catch the base-class exception `RequestException `, which will handle all cases or you can catch them separately and do different things

Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Thanks!

File: openedx/core/djangoapps/edx_global_analytics/tasks.py (hunk 18 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Yes, done.

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 19]
```diff
@@ -3,70 +3,79 @@
...
+
+    active_students_amount_month = fetch_instance_information(
+        'active_students_amount_month', 'active_students_amount',
+        get_previous_month_start_and_end_dates(), cache_timeout_month()
+    )
+
+    return active_students_amount_day, active_students_amount_week, active_students_amount_month
+
+
+def enthusiast_level_statistics_bunch():
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Add docstring

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 20]
```diff
@@ -3,70 +3,79 @@
...
+    students_per_country = fetch_instance_information(
+        'students_per_country', 'students_per_country',
+        get_previous_day_start_and_end_dates(), cache_timeout=None
+    )
+
+    return students_per_country
+
 
 @task
 def count_data():
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Better to name it `collect_stats`

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 21]
```diff
@@ -3,70 +3,79 @@
...
-    geographical coordinates of the platform, courses amount and
-    makes a POST request with the data to the appropriate service.
+    Periodic task function that gathers instance information as like as platform name,
+    active students amount, courses amount etc., then makes a POST request with the data to the appropriate service.
+
+    Sending information depends on statistics level in settings, that have an effect on bunch of data size.
     """
 
     # OLGA settings
     olga_settings = settings.ENV_TOKENS.get('OPENEDX_LEARNERS_GLOBAL_ANALYTICS')
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: What if settings are not set in config file? Do you handle this situation or task will crash?

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 22]
```diff
@@ -80,31 +89,59 @@ def count_data():
...
+    # Enthusiast level (extend Paranoid level)
     if statistics_level == 1:
+        # Get IP address of the platform and convert it to latitude, longitude.
+        ip_data = requests.get('http://freegeoip.net/json')
+        ip_data_json = json.loads(ip_data.text)
+
+        platform_latitude = olga_settings.get("PLATFORM_LATITUDE")
+        platform_longitude = olga_settings.get("PLATFORM_LONGITUDE")
+
+        if platform_latitude and platform_longitude:
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Instead of this `if`, add `ip_data_json['latitude']` and `ip_data_json['longitude']` as a `default` param to `olga_settings.get()`

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 23]
```diff
@@ -80,31 +89,59 @@ def count_data():
...
+            'students_per_country': json.dumps(students_per_country)
         })
 
-    requests.post(post_url, data)
+    try:
+        request = requests.post(post_url, data)
+        logger.info('Connected without error to {0}'.format(request.url))
+        logger.info('Data was successfully sent. Status code is {0}.'.format(request.status_code))
+    except requests.ConnectionError as error:
+        logger.info(error.message)
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Use `logger.exception` to log an exception itself, not just a message

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+    def active_students_amount_default_data():
+        """
+        Default integration and particular data for active students amount functionality.
+        """
+        users_last_login = [
+            timezone.make_aware(datetime.datetime(2017, 5, 14, 23, 59, 59), timezone.get_default_timezone()),
+            timezone.make_aware(datetime.datetime(2017, 5, 15, 00, 00, 00), timezone.get_default_timezone()),
+            timezone.make_aware(datetime.datetime(2017, 5, 15, 23, 59, 59), timezone.get_default_timezone()),
+            timezone.make_aware(datetime.datetime(2017, 5, 16, 00, 00, 00), timezone.get_default_timezone()),
+            timezone.make_aware(datetime.datetime(2017, 5, 16, 00, 00, 01), timezone.get_default_timezone())
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: Why do you use `00`, `01` and so on? Regular `0`, `1` are good too.

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+
+from ..utils import fetch_instance_information, cache_instance_data
+
+
+class TestStudentsAmountPerParticularPeriod(TestCase):
+    """
+    Tests cover all methods, that have a deal with statistics calculation.
+    """
+    @staticmethod
+    def active_students_amount_default_data():
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: It's considered a bad practice to plant a side-effect into a function. See https://stackoverflow.com/a/766736/231412

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py [hunk 26]
```diff
@@ -0,0 +1,157 @@
...
+
+    def test_miss_statistics_query_fetch_instance_information_for_active_students_amount(self):
+        """
+        Verifies that fetch_instance_information raise `KeyError` if default statistics queries don't have
+        corresponding name (dict`s key).
+        """
+        activity_period = datetime.date(2017, 5, 15), datetime.date(2017, 5, 16)
+        cache_timeout = None
+
+        self.assertRaises(KeyError, lambda: fetch_instance_information(
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: Why not use `assertRaises` as a context manager and drop `lambda`?

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py [hunk 27]
```diff
@@ -0,0 +1,157 @@
...
+    def test_fetch_instance_information_for_students_per_country(self):
+        """
+        Verifies that students_per_country returns data as expected in particular period and accurate datetime.
+        """
+        last_login = timezone.make_aware(datetime.datetime(2017, 5, 15, 14, 23, 23), timezone.get_default_timezone())
+        countries = [u'US', u'CA']
+
+        for country in countries:
+            self.user = UserFactory.create(last_login=last_login)
+            self.profile = self.user.profile
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: Why do you assign `self.user` and `self.profile` if you don't use it later?

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py [hunk 28]
```diff
@@ -0,0 +1,157 @@
...
+        self.assertItemsEqual(result, {u'US': 1, u'CA': 1})
+
+    def test_no_students_with_country_fetch_instance_information_for_students_per_country(self):
+        """
+        Verifies that students_per_country returns data as expected if no students with country.
+        """
+        last_login = timezone.make_aware(datetime.datetime(2017, 5, 15, 14, 23, 23), timezone.get_default_timezone())
+
+        for country in range(2):
+            self.user = UserFactory.create(last_login=last_login)
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: Why assign to `self.user`?

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py [hunk 29]
```diff
@@ -0,0 +1,157 @@
...
+
+        self.assertItemsEqual(result, {u'US': 1, u'CA': 1})
+
+    def test_no_students_with_country_fetch_instance_information_for_students_per_country(self):
+        """
+        Verifies that students_per_country returns data as expected if no students with country.
+        """
+        last_login = timezone.make_aware(datetime.datetime(2017, 5, 15, 14, 23, 23), timezone.get_default_timezone())
+
+        for country in range(2):
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: Why do you need 2 countries in this test?

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py [hunk 30]
```diff
@@ -0,0 +1,157 @@
...
+
+        self.assertEqual(result, {None: 0})
+
+
+class TestCacheInstanceData(TestCase):
+    """
+    Tests cover cache-functionality for queries results.
+    """
+    @staticmethod
+    def cache_instance_data_default_data():
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: If this method is the same as in previous Test class -- extract it to the base class.
See https://refactoring.guru/extract-superclass

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py [hunk 31]
```diff
@@ -0,0 +1,157 @@
...
+        """
+        Default integration and particular data for cache instance information tests.
+        """
+        users_last_login = [
+            timezone.make_aware(datetime.datetime(2017, 5, 8, 00, 00, 00), timezone.get_default_timezone()),
+            timezone.make_aware(datetime.datetime(2017, 5, 14, 23, 59, 59), timezone.get_default_timezone()),
+            timezone.make_aware(datetime.datetime(2017, 5, 15, 00, 00, 00), timezone.get_default_timezone()),
+            timezone.make_aware(datetime.datetime(2017, 5, 15, 00, 00, 01), timezone.get_default_timezone())
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: Same thing about `01` vs `1`

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py [hunk 32]
```diff
@@ -0,0 +1,157 @@
...
+        for user_last_login in users_last_login:
+            UserFactory(last_login=user_last_login)
+
+        period_start, period_end = datetime.date(2017, 5, 8), datetime.date(2017, 5, 15)
+
+        return period_start, period_end
+
+    def test_cache_instance_data(self):
+        """
+        Verifies that cache_instance_dat returns data as expected after caching it.
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: Typo `cache_instance_dat`

File: openedx/core/djangoapps/edx_global_analytics/utils.py [hunk 33]
```diff
@@ -0,0 +1,142 @@
+"""
+Helpers for the edX global analytics application.
+"""
+
+import datetime
+import calendar
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: Plese sort alphabetically

File: openedx/core/djangoapps/edx_global_analytics/utils.py [hunk 34]
```diff
@@ -0,0 +1,142 @@
...
+    statistics_queries = {
+        'active_students_amount': UserProfile.objects.exclude(
+            Q(user__last_login=None) | Q(user__is_active=False)
+        ).filter(user__last_login__gte=period_start, user__last_login__lt=period_end).count(),
+
+        'students_per_country': dict(UserProfile.objects.exclude(
+            Q(user__last_login=None) | Q(user__is_active=False)
+        ).filter(user__last_login__gte=period_start, user__last_login__lt=period_end).values(
+            'country'
+        ).annotate(count=Count('country')).values_list('country', 'count'))
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: Change indentation according to [Open edX Python code style guide](https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python_guidelines.html).

File: openedx/core/djangoapps/edx_global_analytics/utils.py [hunk 35]
```diff
@@ -0,0 +1,142 @@
...
+    if cache_timeout is not None:
+        return cache_instance_data(name_to_cache, statistics_queries[query_type], cache_timeout)
+
+    return statistics_queries[query_type]
+
+
+def cache_instance_data(name_to_cache, query_result, cache_timeout):
+    """
+     Caches queries, that calculate particular instance data,
+     including long time unchangeable as like as monthly statistics.
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: `as like as` looks wrong.

File: openedx/core/djangoapps/edx_global_analytics/views.py [hunk 36]
```diff
@@ -15,25 +15,24 @@ class ReceiveTokenView(View):
...
         """
 
         try:
-            received_data = self.request.POST
+            received_data = request.POST
             secret_token = str(received_data.get('secret_token'))
             TokenStorage.objects.update_or_create(
                 pk=1, defaults={"secret_token": secret_token})
-            return HttpResponse(status=200)
+            return HttpResponse(status=201)
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/views.py]: Why this change? It's a misleading detail.

File: openedx/core/djangoapps/edx_global_analytics/views.py [hunk 37]
```diff
@@ -15,25 +15,24 @@ class ReceiveTokenView(View):
...
         except ValueError:
-            raise Http404()
+            return HttpResponse(status=400)
```
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/views.py]: Why this change?

File: openedx/core/djangoapps/edx_global_analytics/tasks.py (hunk 21 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Thanks, I`ll log it.

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py (hunk 26 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: I tried to do as you said, but after asserting I got error, that I don't need.

File: openedx/core/djangoapps/edx_global_analytics/views.py (hunk 36 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/views.py]: On my opinion 201 more expressive. I din't any strict rules about that, but found some examples, where successful token creation returns status 201.

File: openedx/core/djangoapps/edx_global_analytics/views.py (hunk 37 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/views.py]: I think status 404 better says, that a client have issue with the request.

File: openedx/core/djangoapps/edx_global_analytics/utils.py (hunk 35 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: Thanks.

File: openedx/core/djangoapps/edx_global_analytics/utils.py (hunk 33 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: My bad, thanks.

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py (hunk 32 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: Thanks.

File: openedx/core/djangoapps/edx_global_analytics/tasks.py (hunk 23 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: You are right. Done.

File: openedx/core/djangoapps/edx_global_analytics/tasks.py (hunk 20 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Ok, thanks.

File: openedx/core/djangoapps/edx_global_analytics/tasks.py [hunk 38]
```diff
@@ -3,70 +3,79 @@
...
+    get_previous_month_start_and_end_dates,
+    cache_timeout_week,
+    cache_timeout_month
+)
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+def paranoid_level_statistics_bunch():
```
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Done.

File: openedx/core/djangoapps/edx_global_analytics/tasks.py (hunk 19 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: Done.

File: openedx/core/djangoapps/edx_global_analytics/views.py (hunk 37 above)
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/views.py]: `Http404` is returning the same 400 status, but also uses common template for errors.

File: openedx/core/djangoapps/edx_global_analytics/views.py (hunk 36 above)
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/views.py]: Yes, it's more expressive. What bothers me is that this function does not always create new object. So returning 201 in a case of the update is misleading and useless.

Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/views.py]: I found this function does not update token, it always exists or doesn't. 
I`ll rewrite this logic, ticket already in Trello, we just was focused on another logic.

File: openedx/core/djangoapps/edx_global_analytics/views.py (hunk 37 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/views.py]: Now I`ll return that, but it`s need to be discussed (return`s conformity). Thanks.

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py (hunk 29 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: No reason to create few students without country, okay to do with one.

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py (hunk 28 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: Sorry, just `copypaste` from edX examples.

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py (hunk 27 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: The same below.

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py (hunk 24 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: Naive think about precision.

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py (hunk 31 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: Naive think about precision.

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py (hunk 25 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: Done.

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py (hunk 30 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: Thanks.

File: openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py (hunk 24 above)
Comment (z4y4ts) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: Leading zeros don't give you any precision.
What's worse, leading zeros actually give you octal numbers. Not a disaster in this particular case, but certainly not what you intend.

Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tests/test_utils.py]: I understood, i did it correctly as you said.

File: openedx/core/djangoapps/edx_global_analytics/utils.py (hunk 34 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/utils.py]: Done.

File: openedx/core/djangoapps/edx_global_analytics/tasks.py (hunk 22 above)
Comment (DmytryiStriletskyi) [inline openedx/core/djangoapps/edx_global_analytics/tasks.py]: We will change latitude and longitude logic. Just set city instead of ip`s.
Ticket already in Trello, thanks and sorry.
